Utils/Dataflow/009_oracleConnector/Oracle2JSON.py

```diff
@@ -91,25 +90,57 @@ def read_config(config_file):
                          % (config_file, e))
         return None
 
-    try:
-        offset_file = config.get("logging", "offset_file")
-    except (ConfigParser.NoSectionError, ConfigParser.NoOptionError):
-        sys.stderr.write('(INFO) Config: "offset_file" not found in section'
-                         ' "logging". Using default value.\n')
-        offset_file = '.offset'
-    result['offset_file'] = config_path(offset_file, result)
+    # Optional parameters
+    result['final_date'] = str2date(config_get(config, "timestamps", "final"))
+    result['initial_date'] = str2date(
+        config_get(config, "timestamps", "initial",
+                   datetime.utcfromtimestamp(0)))
+    result['offset_file'] = config_path(config_get(config, "logging",
+                                                   "offset_file", ".offset"),
+                                        result)
+
+    return result
 
-    try:
-        initial_date = str2date(config.get("timestamps", "initial"))
-    except (ConfigParser.NoSectionError, ConfigParser.NoOptionError):
-        initial_date = None
 
-    if initial_date is None:
-        sys.stderr.write('(INFO) Config: "initial" date (section'
-                         ' "timestamps") is not configured. Using default'
-                         ' value.\n')
-        initial_date = datetime.utcfromtimestamp(0)
-    result['initial_date'] = initial_date
+def config_get(config, section, param, default=None):
+    """ Get $param value from $section of $config with $default value.
+
+    If the parameter (or even section) is not presented in the config,
+      returns default value instead of throwing an exception (as
+      `config.get()` does).
+
+    :param config: config to read from
+    :param section: config section name
+    :param param: config parameter name
+    :param default: default value
+    :type config: ConfigParser object
+    :type section: string
+    :type param: string
+    :param default: object (any type)
+    :return: param value from config or default value
+    :rtype: object
+    """
+    if not isinstance(config, ConfigParser.ConfigParser):
+        raise TypeError("config_get() expects first parameter to be"
+                        " an instance of 'ConfigParser' (get '%s')."
+                        % config.__class__.__name__)
+    if type(section) != str and type(section) != unicode:
+        raise TypeError("config_get() expects second parameter to be"
```

Comment:
What if in `if..else` will be `or` instead of `and`? Can't catch the main idea of this conditional operator. What will be if just a single condition will satisfy?

Reply:
Let me think...
I believe it was an attempt to pass over the fact that Python 3 doesn`t have `unicode` type, but a very lazy one. Or, at least, not very accurate.

The idea is that when `section` is of type `str` (works for both Python 2 and 3), the `if` check won\`t reach the second condition and we won\`t see this message: `NameError: name 'unicode' is not defined`. When it is *not* `str`, in v.2 we simply check if it is `unicode` or not, while in v.3 we have `NameError` exception... not excactly charactonym one (as we\`d expect to see `TypeError` with description, which is within the `if` clause), yet the under same conditions.

If there was `or` instead of `and`, we would never pass the check:
```
>>> for section in ('string', u'string', 123):
...   if type(section) != str or type(section) != unicode:
...     print "Not passed!"
... 
Not passed!
Not passed!
Not passed!
```

Will add `try/except` handler around the construction (or, better, check for the Python version, for readability).

But later we need to add type-check function for input parameters in pyDKB, I think... something like `check_types((set, of, parameters), (set, of, types))`, where "type" can be any Python type / class, and for strings -- `str`, `unicode`, or some inner pyDKB constant named `STRING`.
It would be better then checking the parameters by hand every now and then...

Thank you!

Reply:
Forget to update this discussion...
Done (by adding the Python version check).